local.c

```diff
@@ -1541,6 +1579,92 @@ int local_dequeue_block(struct iio_block_pdata *pdata, bool nonblock)
 	return -ENOSYS;
 }
 
+static struct iio_event_stream_pdata *
+local_open_events_fd(const struct iio_device *dev)
+{
+	struct iio_event_stream_pdata *pdata;
+	int err;
+
+	if (dev->pdata->fd < 0)
+		return iio_ptr(dev->pdata->fd);
+
+	pdata = zalloc(sizeof(*pdata));
+	if (!pdata)
+		return iio_ptr(-ENOMEM);
+
+	pdata->cancel_fd = eventfd(0, EFD_CLOEXEC | EFD_NONBLOCK);
+	if (pdata->cancel_fd == -1) {
+		err = -errno;
+		free(pdata);
+		return iio_ptr(err);
+	}
+
+	err = ioctl_nointr(dev->pdata->fd, IIO_GET_EVENT_FD_IOCTL, &pdata->fd);
+	if (err < 0) {
+		close(pdata->cancel_fd);
+		free(pdata);
+		return iio_ptr(err);
+	}
+
+	return pdata;
+}
+
+static void local_close_events_fd(struct iio_event_stream_pdata *pdata)
+{
+	uint64_t event = 1;
+	int ret;
+
+	ret = write(pdata->cancel_fd, &event, sizeof(event));
+	close(pdata->cancel_fd);
+
+	close(pdata->fd);
+	free(pdata);
+
+	/* Ignore compiler warnings */
+	(void) ret;
+}
+
+static int local_read_event(struct iio_event_stream_pdata *pdata,
+			    struct iio_event *out_event, bool nonblock)
+{
+	struct pollfd pollfd[2] = {
+		{
+			.fd = pdata->fd,
+			.events = POLLIN,
+		}, {
+			.fd = pdata->cancel_fd,
+			.events = POLLIN,
+		}
+	};
+	int ret, timeout_rel = nonblock ? 0 : -1;
+
+	do {
+		ret = poll(pollfd, 2, timeout_rel);
+	} while (ret == -1 && errno == EINTR);
+
+	if ((pollfd[1].revents & (POLLIN | POLLNVAL)))
```

Comment:
Should this be moved after the `ret` checks below?

Reply:
Also, don't we need to read from the eventfd to reset it?

Reply:
We don't actually want to reset it; otherwise the `local_read_event` function would only fail the first time.

As for the order - the `pollfd[1].revents` check is done before the check for `ret < 0` just to be consistent with what's already in `buffer_check_ready`. It shouldn't matter much anyway.

Reply:
Looking back, I see where the docs say that you have to call `iio_event_stream_destroy()` to cancel the read, so makes sense now that there is no reset to read again..